### PR TITLE
enhance switch with icon

### PR DESCRIPTION
**Pull request by finalight**
ezgif-1-c366afcddb

### CODE REVIEW

File: src/@next/Switch/Switch.tsx [hunk 1]
```diff
@@ -1,19 +1,49 @@
...
   checked?: boolean;
   disabled?: boolean;
   onChange: () => void;
   value: string;
+  withIcon?: boolean;
+}
+
+export interface SwitchStyleProps
+  extends React.InputHTMLAttributes<HTMLDivElement> {
+  withIcon: boolean;
```
Comment (ganiirsyadi) [inline src/@next/Switch/Switch.tsx]: use attribute `data-with-icon`

File: src/@next/Switch/SwitchStyle.tsx [hunk 2]
```diff
@@ -54,9 +76,25 @@ export const InputStyle = styled.input`
...
 
     &:before {
-      transform: translate(16px, -50%);
+      ${({ withIcon }: SwitchProps) => ({
+        transform: `translate(${withIcon ? '24px' : Spacing.space16} , -50%);`,
+      })}
     }
   }
+
+  & + ${SwitchStyle} svg:nth-child(1) {
```
Comment (ganiirsyadi) [inline src/@next/Switch/SwitchStyle.tsx:95]: Will it support custom Icon?

File: src/@next/Switch/SwitchStyle.tsx [hunk 3]
```diff
@@ -1,11 +1,31 @@
 import styled from 'styled-components';
-import { Breakpoints, Spacing } from '..';
+import { Breakpoints, Icon, Spacing } from '..';
 
 import { Blue, Neutral } from '../utilities/colors';
+import { Switch, SwitchProps, SwitchStyleProps } from './Switch';
```
Comment (ganiirsyadi) [inline src/@next/Switch/SwitchStyle.tsx]: Remove unused

**Review by ganiirsyadi — CHANGES_REQUESTED** [review]

File: src/@next/Switch/SwitchStyle.tsx (hunk 2 above)
Comment (finalight) [reply src/@next/Switch/SwitchStyle.tsx:95]: nope, only the tick and the cross

File: src/@next/Switch/SwitchStyle.tsx (hunk 3 above)
Comment (finalight) [reply src/@next/Switch/SwitchStyle.tsx]: why the linting not working!?!?

File: src/@next/Switch/SwitchStyle.tsx [hunk 4]
```diff
@@ -1,11 +1,31 @@
...
+
+export const IconStyle = styled(Icon)`
+  height: 16px;
+
+  fill: ${({ fill }) => fill};
```
Comment (marthayaputra) [inline src/@next/Switch/SwitchStyle.tsx]: can we use `&[data-with-icon] svg { fill: }` to style this on SwitchStyle?, That way you don't need to create IconStyle

File: src/@next/Switch/SwitchStyle.tsx [hunk 5]
```diff
@@ -19,7 +39,9 @@ export const SwitchStyle = styled.div`
   transition: 300ms all;
 
   @media (max-width: ${Breakpoints.large}) {
-    width: 38px;
+    ${({ withIcon }: SwitchStyleProps) => ({
```
Comment (marthayaputra) [inline src/@next/Switch/SwitchStyle.tsx]: same here we can bind a `data-with-icon` attribute here and style this accordingly

File: src/@next/Switch/SwitchStyle.tsx [hunk 6]
```diff
@@ -54,9 +76,25 @@ export const InputStyle = styled.input`
     }
 
     &:before {
-      transform: translate(16px, -50%);
+      ${({ withIcon }: SwitchProps) => ({
```
Comment (marthayaputra) [inline src/@next/Switch/SwitchStyle.tsx]: here as well, we can rely on `data-with-icon` attribute

File: src/@next/Switch/SwitchStyle.tsx [hunk 7]
```diff
@@ -54,9 +76,25 @@ export const InputStyle = styled.input`
...
+        transform: `translate(${withIcon ? '24px' : Spacing.space16} , -50%);`,
```
Comment (marthayaputra) [inline src/@next/Switch/SwitchStyle.tsx]: same as above, give it a default transformation, and override with `&[data-with-icon]`

**Comment by marthayaputra** [comment]
@finalight don't forget to add tests

File: src/@next/Switch/Switch.tsx (hunk 1 above)
Comment (finalight) [reply src/@next/Switch/Switch.tsx]: done

File: src/@next/Switch/SwitchStyle.tsx (hunk 3 above)
Comment (finalight) [reply src/@next/Switch/SwitchStyle.tsx]: done

File: src/@next/Switch/SwitchStyle.tsx (hunk 4 above)
Comment (finalight) [reply src/@next/Switch/SwitchStyle.tsx]: done

File: src/@next/Switch/SwitchStyle.tsx (hunk 5 above)
Comment (finalight) [reply src/@next/Switch/SwitchStyle.tsx]: done

File: src/@next/Switch/SwitchStyle.tsx (hunk 6 above)
Comment (finalight) [reply src/@next/Switch/SwitchStyle.tsx]: done

File: src/@next/Switch/SwitchStyle.tsx (hunk 7 above)
Comment (finalight) [reply src/@next/Switch/SwitchStyle.tsx]: done

File: src/@next/Switch/Switch.stories.tsx [hunk 8]
```diff
@@ -29,6 +30,28 @@ const Template: Story = args => {
...
+const SwitchWithIconTemplate: Story = args => {
+  const [checked, setChecked] = useState(args.checked);
+
+  return (
+    <>
+      <div>The checkbox checked is {checked ? 'true' : 'false'}</div>
+      <div>
+        <Switch
+          {...args}
+          value={'switch'}
```
Comment (ganiirsyadi) [inline src/@next/Switch/Switch.stories.tsx:57]: The previous example you passed `checked` to `<Switch>`. Any concern why you don't pass it here?

File: src/@next/Switch/Switch.tsx [hunk 9]
```diff
@@ -7,13 +9,36 @@ export interface SwitchProps
...
+    const IconRender = () => {
+      return (
+        <>
+          <Icon name="ri-check" fill={Neutral.B100} />
+          <Icon name="ri-close" fill={Neutral.B100} />
+        </>
+      );
+    };
+
+    console.log({ checked });
```
Comment (ganiirsyadi) [inline src/@next/Switch/Switch.tsx]: remove

File: src/@next/Switch/Switch.tsx [hunk 10]
```diff
@@ -24,9 +49,12 @@ export const Switch = React.forwardRef<HTMLInputElement, SwitchProps>(
           type="checkbox"
           onChange={onChange}
           ref={ref}
+          withIcon={withIcon}
         />
 
-        <SwitchStyle data-disabled={disabled} />
+        <SwitchStyle data-disabled={disabled}>
```
Comment (ganiirsyadi) [inline src/@next/Switch/Switch.tsx]: Should pass `data-withicon` here?

File: src/@next/Switch/Switch.tsx [hunk 11]
```diff
@@ -7,13 +9,35 @@ export interface SwitchProps
   disabled?: boolean;
   onChange: () => void;
   value: string;
+  withIcon?: boolean;
+}
+
+export interface IconStyleProps
```
Comment (marthayaputra) [inline src/@next/Switch/Switch.tsx]: Do we still need this?

File: src/@next/Switch/SwitchStyle.tsx [hunk 12]
```diff
@@ -42,7 +67,7 @@ export const SwitchStyle = styled.div`
   }
 `;
 
-export const InputStyle = styled.input`
+export const InputStyle = styled.input<SwitchProps>`
```
Comment (ganiirsyadi) [inline src/@next/Switch/SwitchStyle.tsx]: I dont think SwitchProps will be needed anymore

**Review by ganiirsyadi — CHANGES_REQUESTED** [review]

File: src/@next/Switch/Switch.stories.tsx (hunk 8 above)
Comment (finalight) [reply src/@next/Switch/Switch.stories.tsx:45]: > @finalight don't forget to add tests

added

**Comment by finalight** [comment]
> @finalight don't forget to add tests

added

File: src/@next/Switch/SwitchStyle.tsx (hunk 12 above)
Comment (finalight) [reply src/@next/Switch/SwitchStyle.tsx]: removed

File: src/@next/Switch/Switch.tsx (hunk 11 above)
Comment (finalight) [reply src/@next/Switch/Switch.tsx]: removed

File: src/@next/Switch/Switch.tsx (hunk 10 above)
Comment (finalight) [reply src/@next/Switch/Switch.tsx]: it's passed

File: src/@next/Switch/Switch.tsx (hunk 9 above)
Comment (finalight) [reply src/@next/Switch/Switch.tsx]: removed

File: src/@next/Switch/Switch.stories.tsx (hunk 8 above)
Comment (finalight) [reply src/@next/Switch/Switch.stories.tsx:57]: removed the `checked={checked}` for both story

**Review by marthayaputra — APPROVED** [review]

**Review by ganiirsyadi — APPROVED** [review]

**Review by ganiirsyadi — APPROVED** [review]

**Review by chiahou — APPROVED** [review]
QC passed

**Pull request merged**